src/adapter/src/optimize/dataflow_expiration.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! Helper function for dataflow expiration checks.
+//!
+//! A [`TimeDependence`] describes how a dataflow follows wall-clock time, and this module offers
+//! functions to compute the time dependence based on the current state of the catalog.
+//!
+//! For a set of global IDs, the helper determines how a dataflow that depends on the set of IDs
+//! follows wall-clock time. Within the catalog, we remember how all installed objects follow
+//! wall-clock time, which turns this step into gathering and combining different time dependencies:
+//! * A meet of anything with wall-clock time results in wall-clock time.
+//! * A meet of anything but wall-clock time and a refresh schedule results in a refresh schedule
+//!   that depends on the deduplicated collection of dependencies.
+//! * Otherwise, a dataflow is indeterminate, which expresses that we either don't know how it
+//!   follows wall-clock time, or is a constant collection.
```

Comment:
You could add a note here that if there are both determinate and indeterminate `TimeDependence`s, then it's ok whatever the actual, real time dependence of the indeterminate ones are:
- It's ok if the indeterminate ones have a later frontier than the determinate ones, because then the frontier will be determined by the determinate ones.
- It's also ok if the indeterminate ones have an earlier frontier than the determinate ones. This will mean that the dataflow will expire later than ideal, but it won't cause a panic or wrong results.